internal/debug/debug.go

```diff
@@ -0,0 +1,165 @@
+package debug
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os/exec"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type DebugPod struct {
+	Name      string
+	Namespace string
+	NodeName  string
+}
+
+const (
+	interval       = time.Millisecond * 500
+	duration       = time.Second * 5
+	debugNamespace = "openshift-debug"
+)
+
+// New creates debug pod on the given node, puts it in infinite sleep,
+// and returns the DebugPod object. Use the Clean() method to delete it.
+func New(node string) (*DebugPod, error) {
+	errCh := make(chan error)
+	go func() {
+		errCh <- start(node)
+	}()
+	for {
+		select {
+		case err := <-errCh:
+			return nil, err
+		default:
+			return waitForPod()
+		}
+	}
+}
+
+func (dp *DebugPod) Exec(cmd string) ([]byte, error) {
+	cmdOnDebugPod := append([]string{"exec", "-n", dp.Namespace, dp.Name, "--", "chroot", "/host"}, strings.Split(cmd, " ")...)
+	out, err := exec.Command("oc", cmdOnDebugPod...).CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("failed to exec command \"%s\" on node %s: %v\n%s", cmd, dp.NodeName, err, string(out))
+	}
+
+	return out, nil
+}
+
+func (dp *DebugPod) ExecWithRetry(cmd string) ([]byte, error) {
+	var (
+		out     []byte
+		execErr error
+	)
+
+	if err := wait.PollUntilContextTimeout(context.TODO(), interval, duration, true, func(ctx context.Context) (bool, error) {
+		out, execErr = dp.Exec(cmd)
+		if execErr != nil {
+			return false, execErr
+		}
+
+		return true, execErr
+	}); err != nil {
+		return nil, errors.Join(err, execErr)
+	}
+
+	return out, nil
+}
+
+func (dp *DebugPod) Clean() error {
+	output, err := exec.Command("oc", "delete", "pod", "-n", dp.Namespace, dp.Name).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed deleting debug pod %s/%s: %v\n%s", dp.Namespace, dp.Name, err, string(output))
+	}
+
+	output, err = exec.Command("oc", "delete", "ns", dp.Namespace).CombinedOutput()
```

Comment:
why delete the entire namespace? (does each pod gets its own ns?)

Reply:
yes

Reply:
however, maybe it's better to set the namespace myself
because I found out that if the "oc" tool set the working project to something,
it will create them in the current project. and that will fail everything.